[JSC] Watchdog calls valid timers "stale" on Windows

rkirsling opened the pull request:
#### 512db9a6900b400f9ce93e90e6e178fa74fef03e
<pre>
[JSC] Watchdog calls valid timers &quot;stale&quot; on Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=255135">https://bugs.webkit.org/show_bug.cgi?id=255135</a>

Reviewed by Mark Lam.

The following tests tend to time out on Win JSC:
- stress/bounds-checking-in-cold-loop.js
- stress/call-link-info-osrexit-repatch.js
- stress/call-varargs-inlining-should-not-clobber-previous-to-free-register.js
- stress/ftl-ai-filter-phantoms-should-clear-clear-value.js
- stress/ftl-entry-osr-exit-has-nullptr-node.js
- stress/has-indexed-property-with-worsening-array-mode.js
- stress/polling-based-trap-on-unlinked-dfg.js
- stress/reduce-loop-strength-multiple-exits.js
- stress/regress-189227-watchdog-on-infinite-loop.js
- stress/suppress-TerminationException-in-operationGetPrivateNameOptimize.js

The cause is that Watchdog::shouldTerminate fires as much as 15ms prior to m_deadline,
so one simple solution is to add a 20ms epsilon value when comparing &quot;now&quot; against the deadline.

Technically, we wouldn&apos;t need to platform-restrict this value -- after all, the early out appears more concerned with
&quot;when m_deadline is infinity&quot; than &quot;when we&apos;re a little early&quot; -- but then, the issue only seems to occur on Windows.

* Source/JavaScriptCore/runtime/Watchdog.cpp:
(JSC::Watchdog::shouldTerminate):

Canonical link: <a href="https://commits.webkit.org/262701@main">https://commits.webkit.org/262701@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03c793bb04f496262723a17bdf43e7ee33d143e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2313 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/2325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3257 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2340 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2417 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2399 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2080 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/2038 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2083 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3134 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/25 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2068 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1943 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1939 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2143 "Built successfully") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/2065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3256 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2221 "Built successfully and passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2116 "Passed tests") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2381 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2120 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2074 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/535 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/574 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2051 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2432 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/261 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2243 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/649 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/Watchdog.cpp:
```diff
@@ -60,7 +60,9 @@ void Watchdog::setTimeLimit(Seconds limit,
-    if (MonotonicTime::now() < m_deadline)
+
+    auto epsilon = Seconds::fromMilliseconds(20);
+    if (MonotonicTime::now() + epsilon < m_deadline)
```
I think the use of an epsilon here deserves a bit of explanation for future reference.  Can you add a comment before landing please?

The tests in `testapi` are a bit fine tuned to minimize running time.  I think a 20ms might not matter, but I'm really not sure, and I wouldn't want `testapi` to become more flaky.  I would prefer that you conditionalize this change on `OS(WINDOWS)`, just so I know I won't have to debug potential spurious `testapi` failures.

MenloDorian reviewed (commented):
r=me

rkirsling committed:
[JSC] Watchdog calls valid timers "stale" on Windows
https://bugs.webkit.org/show_bug.cgi?id=255135

Reviewed by Mark Lam.

The following tests tend to time out on Win JSC:
- stress/bounds-checking-in-cold-loop.js
- stress/call-link-info-osrexit-repatch.js
- stress/call-varargs-inlining-should-not-clobber-previous-to-free-register.js
- stress/ftl-ai-filter-phantoms-should-clear-clear-value.js
- stress/ftl-entry-osr-exit-has-nullptr-node.js
- stress/has-indexed-property-with-worsening-array-mode.js
- stress/polling-based-trap-on-unlinked-dfg.js
- stress/reduce-loop-strength-multiple-exits.js
- stress/regress-189227-watchdog-on-infinite-loop.js
- stress/suppress-TerminationException-in-operationGetPrivateNameOptimize.js

The cause is that Watchdog::shouldTerminate fires as much as 15ms prior to m_deadline,
so one simple solution is to add a 20ms epsilon value when comparing "now" against the deadline.

Technically, we wouldn't need to platform-restrict this value -- after all, the early out appears more concerned with
"when m_deadline is infinity" than "when we're a little early" -- but then, the issue only seems to occur on Windows.

* Source/JavaScriptCore/runtime/Watchdog.cpp:
(JSC::Watchdog::shouldTerminate):

Canonical link: https://commits.webkit.org/262701@main
